File: backend/src/main/java/com/horain/model/AgentTurn.java
package com.horain.model;

import jakarta.persistence.*;

import java.time.Instant;
import java.util.UUID;

/**
 * One turn of conversation: user message + assistant response and metadata.
 * Used for trace and later extraction into eval candidates.
 */
@Entity
@Table(name = "agent_turn")
public class AgentTurn {

    @Id
    private UUID id;

    @Column(name = "conversation_id", nullable = false)
    private UUID conversationId;

    @Column(name = "turn_index", nullable = false)
    private Integer turnIndex;

    @Column(name = "user_message", nullable = false, columnDefinition = "TEXT")
    private String userMessage;

    @Column(name = "assistant_message", columnDefinition = "TEXT")
    private String assistantMessage;

    @Column(name = "tool_calls_json", columnDefinition = "TEXT")
    private String toolCallsJson;

    @Column(name = "tool_results_json", columnDefinition = "TEXT")
    private String toolResultsJson;

    @Column(name = "ui_payload_json", columnDefinition = "TEXT")
    private String uiPayloadJson;

    @Column(name = "system_prompt_version", length = 50)
    private String systemPromptVersion;

    @Column(name = "model")
    private String model;

    @Column(name = "status", length = 50)
    private String status;

    @Column(name = "history_snapshot_json", columnDefinition = "TEXT")
    private String historySnapshotJson;

    @Column(name = "context_entries_json", columnDefinition = "TEXT")
    private String contextEntriesJson;

    @Column(name = "latency_ms")
    private Long latencyMs;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    public UUID getId() { return id; }
    public void setId(UUID id) { this.id = id; }
    public UUID getConversationId() { return conversationId; }
    public void setConversationId(UUID conversationId) { this.conversationId = conversationId; }
    public Integer getTurnIndex() { return turnIndex; }
    public void setTurnIndex(Integer turnIndex) { this.turnIndex = turnIndex; }
    public String getUserMessage() { return userMessage; }
    public void setUserMessage(String userMessage) { this.userMessage = userMessage; }
    public String getAssistantMessage() { return assistantMessage; }
    public void setAssistantMessage(String assistantMessage) { this.assistantMessage = assistantMessage; }
    public String getToolCallsJson() { return toolCallsJson; }
    public void setToolCallsJson(String toolCallsJson) { this.toolCallsJson = toolCallsJson; }
    public String getToolResultsJson() { return toolResultsJson; }
    public void setToolResultsJson(String toolResultsJson) { this.toolResultsJson = toolResultsJson; }
    public String getUiPayloadJson() { return uiPayloadJson; }
    public void setUiPayloadJson(String uiPayloadJson) { this.uiPayloadJson = uiPayloadJson; }
    public String getSystemPromptVersion() { return systemPromptVersion; }
    public void setSystemPromptVersion(String systemPromptVersion) { this.systemPromptVersion = systemPromptVersion; }
    public String getModel() { return model; }
    public void setModel(String model) { this.model = model; }
    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    public String getHistorySnapshotJson() { return historySnapshotJson; }
    public void setHistorySnapshotJson(String historySnapshotJson) { this.historySnapshotJson = historySnapshotJson; }
    public String getContextEntriesJson() { return contextEntriesJson; }
    public void setContextEntriesJson(String contextEntriesJson) { this.contextEntriesJson = contextEntriesJson; }
    public Long getLatencyMs() { return latencyMs; }
    public void setLatencyMs(Long latencyMs) { this.latencyMs = latencyMs; }
    public Instant getCreatedAt() { return createdAt; }
    public void setCreatedAt(Instant createdAt) { this.createdAt = createdAt; }

    @PrePersist
    protected void onCreate() {
        if (createdAt == null) createdAt = Instant.now();
    }
}

File: backend/src/main/kotlin/com/horain/model/AgentTurn.kt
package com.horain.model

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.PrePersist
import jakarta.persistence.Table
import java.time.Instant
import java.util.UUID

/**
 * One turn of conversation: user message + assistant response and metadata.
 * Used for trace and later extraction into eval candidates.
 */
@Entity
@Table(name = "agent_turn")
class AgentTurn {
    @Id
    var id: UUID? = null

    @Column(name = "conversation_id", nullable = false)
    var conversationId: UUID? = null

    @Column(name = "turn_index", nullable = false)
    var turnIndex: Int? = null

    @Column(name = "user_message", nullable = false, columnDefinition = "TEXT")
    var userMessage: String? = null

    @Column(name = "assistant_message", columnDefinition = "TEXT")
    var assistantMessage: String? = null

    @Column(name = "tool_calls_json", columnDefinition = "TEXT")
    var toolCallsJson: String? = null

    @Column(name = "tool_results_json", columnDefinition = "TEXT")
    var toolResultsJson: String? = null

    @Column(name = "ui_payload_json", columnDefinition = "TEXT")
    var uiPayloadJson: String? = null

    @Column(name = "system_prompt_version", length = 50)
    var systemPromptVersion: String? = null

    @Column(name = "model")
    var model: String? = null

    @Column(name = "status", length = 50)
    var status: String? = null

    @Column(name = "history_snapshot_json", columnDefinition = "TEXT")
    var historySnapshotJson: String? = null

    @Column(name = "context_entries_json", columnDefinition = "TEXT")
    var contextEntriesJson: String? = null

    @Column(name = "latency_ms")
    var latencyMs: Long? = null

    @Column(name = "created_at", nullable = false)
    var createdAt: Instant? = null

    @PrePersist
    protected fun onCreate() {
        if (createdAt == null) createdAt = Instant.now()
    }
}
